types/src/network_address.rs

```diff
@@ -0,0 +1,651 @@
+/// 2. Open a TCP connection to `(addr, 6180)`.
+/// 3. Perform a Noise IK handshake and assume the peer's static pubkey is
+///    `<x25519-pubkey>`. After this step, we will have a secure, authenticated
+///    connection with the peer.
+/// 4. Perform a LibraNet version negotiation handshake (version 1).
+///
+/// One key concept behind `NetworkAddress` is that it is fully self-describing,
+/// which allows us to easily "pre-negotiate" protocols while also allowing for
+/// future upgrades. For example, it is generally unsafe to negotiate a secure
+/// transport in-band. Instead, with `NetworkAddress` we can advertise (via
+/// discovery) the specific secure transport protocol and public key that we
+/// support (and even advertise multiple incompatible versions). When a peer
+/// wishes to establish a connection with us, they already know which secure
+/// transport protocol to use; in this sense, the secure transport protocol is
+/// "pre-negotiated" by the dialier selecting which advertised protocol to use.
+///
+/// In addition, `NetworkAddress` is integrated with the LibraNet concept of a
+/// [`Transport`], which takes a `NetworkAddress` when dialing and peels off
+/// [`Protocols`] to establish a connection and perform initial handshakes.
+/// Similarly, the `Transport` takes `NetworkAddress` to listen on, which tells
+/// it what protocols to expect on the socket.
+///
+/// An example of a serialized `NetworkAddress` and `RawNetworkAddress`:
+///
+/// ```rust
+/// // human-readable format:
+/// //
+/// //   "/ip4/10.0.0.16/tcp/80"
+/// //
+/// // serialized NetworkAddress:
+/// //
+/// //      [ 02 00 0a 00 00 10 05 80 00 ]
+/// //       \  \  \           \  \
+/// //        \  \  \           \  '-- u16 tcp port
+/// //         \  \  \           '-- uvarint protocol id for /tcp
+/// //          \  \  '-- u32 ipv4 address
+/// //           \  '-- uvarint protocol id for /ip4
+/// //            '-- uvarint number of protocols
+/// //
+/// // serialized RawNetworkAddress:
+/// //
+/// //   [ 09 02 00 0a 00 00 10 05 80 00 ]
+/// //    \   \
+/// //     \   '-- serialized NetworkAddress
+/// //      '-- 9 byte uvarint length prefix
+///
+/// use libra_types::network_address::{NetworkAddress, RawNetworkAddress};
+/// use lcs;
+/// use std::{str::FromStr, convert::TryFrom};
+///
+/// let addr = NetworkAddress::from_str("/ip4/10.0.0.16/tcp/80").unwrap();
+/// let raw_addr = RawNetworkAddress::try_from(&addr).unwrap();
+/// let actual_ser_raw_addr = lcs::to_bytes(&raw_addr).unwrap();
+/// let actual_raw_addr: Vec<u8> = raw_addr.into();
+///
+/// let expected_raw_addr: Vec<u8> = [2, 0, 10, 0, 0, 16, 5, 80, 0].to_vec();
+/// let expected_ser_raw_addr: Vec<u8> = [09, 2, 0, 10, 0, 0, 16, 5, 80, 0].to_vec();
+///
+/// assert_eq!(expected_raw_addr, actual_raw_addr);
+/// assert_eq!(expected_ser_raw_addr, actual_ser_raw_addr);
+/// ```
+#[derive(Clone, Eq, PartialEq)]
+pub struct NetworkAddress(Vec<Protocol>);
```

Comment:
it could, but it would significantly increase the parsing complexity : (

instead, I think it's just as good if the `Transport` takes a `NetworkAddress` and decides if it supports it

Reply:
So "/ip6/::1/ip6/::1" will parse successfully but when we try to dial it, the Transport will tell us it's unsupported.

To be honest, I think the suggested approach assumes too much structure about the network addresses and will hinder future changes.

Reply:
Just saw this comment now. Thanks for responding! I think having strict definitions of what can and cannot be valid is really nice. It allows you to evaluate the correctness of an address prior to using it. Considering that we own both pieces of code, it should be pretty easy to move appropriately. The loose definitions allow for a lot of interpretation and flexibility agreed, but I think that defining the specifics of a path prevent confusion and / or abuse. With that being said, I think we can take a let's see what happens approach and re-evaluate.